tutorials/dataframe/df103_NanoAODHiggsAnalysis.C

```diff
@@ -0,0 +1,517 @@
+         return true;
+      }
+      return false;
+   };
+   auto df_pt = df_eta.Filter(pt_cuts, {"Muon_pt", "Electron_pt"}, "Pt cuts");
+   auto dr_cuts = [](rvec_f mu_eta, rvec_f mu_phi, rvec_f el_eta, rvec_f el_phi) {
+      auto mu_dr = sqrt(pow(mu_eta[0] - mu_eta[1], 2) + pow(mu_phi[0] - mu_phi[1], 2));
+      auto el_dr = sqrt(pow(el_eta[0] - el_eta[1], 2) + pow(el_phi[0] - el_phi[1], 2));
+      if (mu_dr < 0.02 || el_dr < 0.02) {
+         return false;
+      }
+      return true;
+   };
+   auto df_dr = df_pt.Filter(dr_cuts, {"Muon_eta", "Muon_phi", "Electron_eta", "Electron_phi"}, "Dr cuts");
+   auto df_iso = df_dr.Filter("All(abs(Electron_pfRelIso03_all)<0.40) && "
+                              "All(abs(Muon_pfRelIso04_all)<0.40)",
+                              "Require good isolation");
+   auto df_el_ip3d = df_iso.Define("Electron_ip3d_el", "sqrt(Electron_dxy*Electron_dxy + Electron_dz*Electron_dz)");
+   auto df_el_sip3d = df_el_ip3d.Define("Electron_sip3d_el", "Electron_ip3d_el/sqrt(Electron_dxyErr*Electron_dxyErr + "
+                                                             "Electron_dzErr*Electron_dzErr)");
+   auto df_el_track = df_el_sip3d.Filter("All(Electron_sip3d_el<4) && All(abs(Electron_dxy)<0.5) && "
+                                         "All(abs(Electron_dz)<1.0)",
+                                         "Electron track close to primary vertex with small uncertainty");
+   auto df_mu_ip3d = df_el_track.Define("Muon_ip3d_mu", "sqrt(Muon_dxy*Muon_dxy + Muon_dz*Muon_dz)");
+   auto df_mu_sip3d = df_mu_ip3d.Define("Muon_sip3d_mu", "Muon_ip3d_mu/sqrt(Muon_dxyErr*Muon_dxyErr + "
+                                                         "Muon_dzErr*Muon_dzErr)");
+   auto df_mu_track = df_mu_sip3d.Filter("All(Muon_sip3d_mu<4) && All(abs(Muon_dxy)<0.5) && "
+                                         "All(abs(Muon_dz)<1.0)",
+                                         "Muon track close to primary vertex with small uncertainty");
+   auto df_2p2n = df_mu_track.Filter("Sum(Electron_charge)==0 && Sum(Muon_charge)==0",
+                                     "Two opposite charged electron and muon pairs");
+   return df_2p2n;
+}
+
+// Reconstruct two Z candidates from four leptons of the same kind
+RVec<RVec<size_t>> reco_zz_to_4l(rvec_f pt, rvec_f eta, rvec_f phi, rvec_f mass, rvec_i charge)
+{
+   RVec<RVec<size_t>> idx(2);
+
+   // Find first lepton pair with invariant mass closest to Z mass
+   const auto z_mass = 91.2;
+   auto idx_cmb = Combinations(pt, 2);
+   auto best_mass = -1;
+   size_t best_i1 = 0;
+   size_t best_i2 = 0;
+   for (size_t i = 0; i < idx_cmb[0].size(); i++) {
+      const auto i1 = idx_cmb[0][i];
+      const auto i2 = idx_cmb[1][i];
+      if (charge[i1] != charge[i2]) {
+         TLorentzVector p1, p2;
+         p1.SetPtEtaPhiM(pt[i1], eta[i1], phi[i1], mass[i1]);
+         p2.SetPtEtaPhiM(pt[i2], eta[i2], phi[i2], mass[i2]);
+         const auto this_mass = (p1 + p2).M();
```

Comment:
Food for thought: we perhaps need something in vecops to avoid this boiler plate